src/lib.rs

```diff
@@ -170,21 +170,33 @@ impl<A: Discard> DiscardOnDrop<A> {
     /// This is implemented as a function (*not* a method) so that way it doesn't interfere with any of the
     /// methods on `discarder`.
     #[inline]
-    pub fn leak(mut this: Self) -> A {
-        match this.0.take() {
-            Some(value) => value,
-            None => unreachable!(),
+    pub fn leak(this: Self) -> A {
+        // We want to move the `A` out of `this`, but that's not allowed because `this` implements `Drop`
+        // (and we must also avoid calling `drop()` on `this` or else `A` would get dropped twice).
+        // We can do that move by using the unsafe function std::ptr::read(),
+        // and then use `mem::forget()` on `this` so it never gets dropped. The `A` will get dropped by the caller.
+        unsafe {
+            let value : A = ::std::ptr::read(this.0.deref());
+            ::std::mem::forget(this);
+            value
         }
     }
 }
 
 impl<A: Discard> Drop for DiscardOnDrop<A> {
     #[inline]
     fn drop(&mut self) {
-        match self.0.take() {
-            Some(discarder) => discarder.discard(),
-            // The discarder was leaked
-            None => {},
+        // This only gets called if there is still a valid `A` inside the `ManuallyDrop`,
+        // since in `leak()` we prevent `drop()` from being called.
+        //
+        // Similar to `leak()`, we want to move `A` out of `self` but again we can't,
+        // this time because we only have a mutable reference, not a value.
+        // The solution is the same though, use `std::ptr::read()` to do the move,
+        // the `A` will get dropped by `.discard()` and since we wrapped it in `ManuallyDrop`,
+        // it won't be dropped again at the end of this function.
+        unsafe {
```

Comment:
Why doesn't this use [`drop`](https://doc.rust-lang.org/std/mem/union.ManuallyDrop.html#method.drop)?

Reply:
`Discard::discard()` takes `self` by value, so the `value` we pass to it will get dropped in that function. If we were to drop it again using `ManuallyDrop::drop`, that would be a double delete.

Reply:
Ah, right, of course, not sure why I didn't realize that.